include/dali/core/convert.h

```diff
@@ -165,14 +189,45 @@ DALI_HOST_DEV constexpr bool clamp(T value, ret_type<bool>) {
   return static_cast<bool>(value);
 }
 
+
+#if defined(__clang__) && defined(__CUDA__)
+
 template <typename T>
-DALI_HOST_DEV constexpr float16 clamp(T value, ret_type<float16>) {
+DALI_HOST constexpr T clamp(float16 value, ret_type<T>) {
+  return clamp(static_cast<float>(value), ret_type<T>());
+}
+
+DALI_HOST constexpr float16 clamp(float16 value, ret_type<float16>) {
+  return value;
+}
+
+template <typename T>
+DALI_DEVICE constexpr T clamp(__half value, ret_type<T>) {
+  return clamp(static_cast<float>(value), ret_type<T>());
+}
+
+DALI_DEVICE constexpr __half clamp(__half value, ret_type<__half>) {
+  return value;
+}
+
+template <typename T>
+DALI_HOST constexpr float16 clamp(T value, ret_type<float16>) {
   constexpr float f16_min = -65504.0f, f16_max = 65504.0f;
   float f = clamp(value, ret_type<float>());
   f = f < f16_min ? f16_min : f > f16_max ? f16_max : f;
   return static_cast<float16>(f);
 }
 
+template <typename T>
+DALI_DEVICE constexpr __half clamp(T value, ret_type<__half>) {
+  constexpr float f16_min = -65504.0f, f16_max = 65504.0f;
```

Comment:
Is there any CUDA define for max/min value of half?

Reply:
Not that I'm aware of.